src/asyncresults.jl

```diff
@@ -0,0 +1,256 @@
+    end
+end
+
+iserror(async_result::AsyncResult) = Base.istaskfailed(async_result.result_task)
+Base.isready(async_result::AsyncResult) = istaskdone(async_result.result_task)
+Base.wait(async_result::AsyncResult) = wait(async_result.result_task)
+Base.fetch(async_result::AsyncResult) = fetch(async_result.result_task)
+trywait(async_result::AsyncResult) = (try wait(async_result) catch end; nothing)
+trywait(::Nothing) = nothing
+Base.close(async_result::AsyncResult) = cancel(async_result)
+
+"""
+    async_execute(
+        jl_conn::Connection,
+        query::AbstractString,
+        [parameters::Union{AbstractVector, Tuple},]
+        kwargs...
+    ) -> AsyncResult
+
+Run a query on the PostgreSQL database and return an [`AsyncResult`](@ref).
+
+The `AsyncResult` contains a `Task` which processes a query asynchronously.
+Calling `fetch` on the `AsyncResult` will return a [`Result`](@ref).
+
+All keyword arguments are the same as [`execute`](@ref) and are passed to the created
+`Result`.
+
+`async_execute` does not yet support [`Statement`](@ref)s.
+
+`async_execute` optionally takes a `parameters` vector which passes query parameters as
+strings to PostgreSQL.
+Queries without parameters can contain multiple SQL statements, and the result of the final
+statement is returned.
+Any errors which occur during executed statements will be bundled together in a
+`CompositeException` and thrown.
+
+As is normal for `Task`s, any exceptions will be thrown when calling `wait` or `fetch`.
+"""
+function async_execute end
+
+function async_execute(jl_conn::Connection, query::AbstractString; kwargs...)
+    async_result = _async_execute(jl_conn; kwargs...) do jl_conn
+        _async_submit(jl_conn.conn, query)
+    end
+
+    return async_result
+end
+
+function async_execute(
+    jl_conn::Connection,
+    query::AbstractString,
+    parameters::Union{AbstractVector, Tuple};
+    kwargs...
+)
+    string_params = string_parameters(parameters)
+    pointer_params = parameter_pointers(string_params)
+
+    async_result = _async_execute(jl_conn; kwargs...) do jl_conn
+        _async_submit(jl_conn.conn, query, pointer_params)
+    end
+
+    return async_result
+end
+
+function _async_execute(
+    submission_fn::Base.Callable, jl_conn::Connection; throw_error::Bool=true, kwargs...
+)
+    async_result = AsyncResult(jl_conn; kwargs...)
+
+    async_result.result_task = @async lock(jl_conn) do
+        jl_conn.async_result = async_result
+
+        try
+            success = submission_fn(jl_conn)
+
+            if !success
+                # does not respect `throw_error` as there's no result to return on this error
+                error(LOGGER, error_message(async_result.jl_conn))
+            end
+
+            result = handle_result(async_result; throw_error=throw_error)
+
+            return result::Result
+        finally
+            jl_conn.async_result = nothing
+        end
+    end
+
+    return async_result
+end
```

Comment:
It is weird that I am saying this, but can we reduce the amount of vertical soace this takes up?
By removing some white-space,
and possibly rmoving some variables that are defined then used just once?

I think getting rid of some of the vertical length would make dealing with the fact that it is so nested easier.